lib/sie/parser/line_parser.rb

```diff
@@ -12,18 +12,34 @@ def parse
         first_token = tokens.shift
 
         entry = Entry.new(first_token.label)
-        entry_type = first_token.entry_type
-
-        entry_type.each_with_index do |entry_type, i|
-          break if i >= tokens.size
-
-          if entry_type.is_a?(Hash)
-            skip_array(tokens, i)
-            next
+        line_entry_type = first_token.entry_type
+
+        ti = ai = 0
+        while ti < tokens.size && ai < line_entry_type.size
+          attr_entry_type = line_entry_type[ai]
+
+          if attr_entry_type.is_a?(Hash)
+            label = attr_entry_type[:name]
+            type = attr_entry_type[:type]
+            entry.attributes[label] ||= []
+
+            ti += 1
+            hash_tokens = []
+            while !tokens[ti].is_a?(Sie::Parser::Tokenizer::EndArrayToken)
+              hash_tokens << tokens[ti].value
+              ti += 1
+            end
+
+            hash_tokens.each_slice(type.size).each do |slice|
+              entry.attributes[label] << Hash[type.zip(slice)]
+            end
           else
-            label = entry_type
-            entry.attributes[label] = tokens[i].value
+            label = attr_entry_type
+            entry.attributes[label] = tokens[ti].value
           end
+
+          ti += 1
+          ai += 1
         end
```

Comment:
Could this be written more functionally? Using .map/select/... family of methods? I can't remember seeing "while" in ruby for ages. Also, it's a very long method. It probably does many things and it's hard to see what it does on a high level. Could maybe even be it's own object.